### DISCUSSION TITLE
Rolling back ignoreSelfIntersections to 6.5.0 behaviour

### DISCUSSION THREAD

**Discussion by smallsaucepan** [Turfjs/turf#2728] [General]
Discussion to prepare for a PR that will revert ignoreSelfIntersections behaviour to what is in 6.5.0. With the release of v7 we seem to have introduced multiple breaking changes to this area and they should be reverted ASAP.

Some questions to answer:

1. What was the ignore self intersections behaviour in 6.5.0?
2. Can we trace to some degree where / why that changed? Not to blame anyone. More to understand how we got here and to ensure we aren't missing something.
3. Roughly what will we have to change as part of a PR?

Covers issues:

#2722
#2707
#2700
#2633
#2585

**Comment by smallsaucepan** [comment]
History as I understand it. Will amend as comments and corrections are suggested.

6.5.0 (mid-2021) shipped with an in house algorithm for line-intersect. There is no mention of self intersections and no way to override the behaviour. The (to be confirmed) belief is this implementation always **ignored** self intersections.

In 2022 line-intersect was rewritten to use the third party sweepline-intersections library. This included the function changing to **recognise** self intersections by default. It's not clear yet why this change was made or if it was intended.

Over the ensuing couple of years the ignoreSelfIntersections option has been added in a few more places. The default value chosen may have been based on an incorrect assumption that recognising self intersections was the default in 6.5.0 because that was what was in the development branch.

Eventually, releasing 7.0.0 (mid-2024) shipped an undocumented breaking change to line-intersect (and other packages).

### Possibly impacted packages:

boolean-crosses
boolean-disjoint
boolean-intersects
line-intersect
line-split

### To do

- [x] Confirm 6.5.0 and 7.1.0 behaviour for each possibly affected package
- [ ] Confirm we've covered all the packages that might be effected
- [ ] Flesh out a PR

### Behaviour matrix - lines

| Package | 6.5.0 | 7.1.0 | dev |
|--|--|--|--|
| boolean-crosses | ignores _false_ | 🟢 ignores |
| boolean-disjoint | ignores _true_ | 🔴 recognises |
| boolean-intersects | ignores _false_ | 🔴 recognises |
| line-intersect | ignores _[ ]_ | 🔴 recognises |
| line-split | ignores _[ ]_ | 🟢 ignores |

![Screenshot 2024-10-08 at 10 38 35 PM](https://github.com/user-attachments/assets/aae23269-c622-48d8-9b31-7013447b96a5)


### Behaviour matrix - polygons

| Package | 6.5.0 | 7.1.0 | dev |
|--|--|--|--|
| boolean-crosses | n/a | n/a |
| boolean-disjoint | ignores _true_ | 🔴 recognises |
| boolean-intersects | ignores _false_ | 🔴 recognises |
| line-intersect | ignores _[ ]_ | 🔴 recognises |
| line-split | n/a | n/a |

![Screenshot 2024-10-08 at 10 33 50 PM](https://github.com/user-attachments/assets/72e1b99d-dbcf-4dd8-8815-71e19d94badb)

**Reply by pm0u** [reply]
To support of the point 6.5.0 _did not_ check self intersections - this is the relevant implementation https://github.com/Turfjs/turf/blob/v6.5.0/packages/turf-line-intersect/index.ts#L69-L83

https://github.com/Turfjs/turf/blob/5375941072b90d489389db22b43bfe809d5e451e/packages/turf-line-intersect/index.ts#L69-L83

My understanding of the above is that one geometry (`line2`) is loaded into an Rbush index (line 70).
Then the other feature (`line1`) is split (line 71) and checked for intersections against the RBush tree (line 72), with the resulting featureCollection of the result for each segment fed into `featureEach`. The actual point of intersection for an intersection pair (components of pair are line2, line1) is then computed on line 74 and added to a feature collection if it is not already computed. This feature collection is returned once all iterations are complete.

I don't see a way that this could have accounted for a self intersection. If separate consumers of this did something like `lineIntersects(line1, line2) || lineIntersects(line1, line1) || lineIntersects(line2, line2)` then those consumers would be checking for a self intersection, but I don't think that check is performed anywhere.

**Comment by smallsaucepan** [comment]
Hi @pm0u. Would you mind taking a look at the above? Can you think of any other packages we should check?

Also, you mention on #2722 there is no workaround for booleanCrosses (agreed). I don't see booleanCrosses misbehaving for the test cases above though. Nor any existing issues along those lines. Do you have a test case where it is being affected by this ignores issue?

**Reply by pm0u** [reply]
Do you have the source for this reproduction? This was the reproduction I created: https://codesandbox.io/p/sandbox/turfjs-7-1-0-s8x273

I think the difference may be that my example is using line-poly crossing check - we may want to add some test cases for that as well.


<img width="1717" alt="image" src="https://github.com/user-attachments/assets/527249e3-8973-4487-8309-f38f31052c44">

**Reply by pm0u** [reply]
Based on the source, the line-line example should (assuming we are looking at the correct geo in the image provided) return as crossing - line-line and line-poly both use the `lineIntersect` package without ignoring self intersections

https://github.com/Turfjs/turf/blob/90e91a3c1727bd93cd2888f27d15ec74d7eec224/packages/turf-boolean-crosses/index.ts#L103-L135

**Reply by pm0u** [reply]
I added a test case to my example of line-line comparison and there does not seem to be an issue in my test case either. I am wondering if this `isPointOnLine` check that happens for line-line is the reason why there is a difference. I think this is somewhat diverging behavior - I would think for consistency it could make sense to perform the same check here, eg convert poly to line and then use `doLineStringsCross` within `doLineStringAndPolygonCross`?

https://github.com/Turfjs/turf/blob/90e91a3c1727bd93cd2888f27d15ec74d7eec224/packages/turf-boolean-crosses/index.ts#L113-L118

**Reply by smallsaucepan** [reply]
Thankyou, that's great. I'd missed those tests. Have added them to mine and confirm seeing the same results.

> the line-line example should [...] return as crossing 

Hmm. line-line only seems to use lineIntersect as a short circuit method. Even though lineIntersect is returning a false positive, the nested loops still check each segment with isPointOnLineSegment.

line-poly on the other hand uses lineIntersect verbatim so I reckon that same false positive (which line-line neutralises by looping all segments as well) is why it's only the *-poly combos are failing. What do you think?

**Comment by smallsaucepan** [comment]
Hi @01100100. Copying you in on this issue because it touches on #2633. It looks like we are going to be changing ignoreSelfIntersections to default to true - Turf wide. Within the context of your PR I think this is ok as that was all about letting people _override_ ignore to true. If true becomes the default they're not affected.

Please have a read of the above and let me know if I've got anything wrong. Thanks!

**Comment by 01100100** [comment]
Hello Mr Saucepan,

Sounds good with me, happy the changes helped.

Best regards,
Dave

**Comment by 01100100** [comment]
If I remember correctly, the default pre 6.5.0 behaviour was that it was ignored, hence false. Just so your aware.

Maybe this change will need some tests updating... :test_tube:

**Reply by smallsaucepan** [reply]
> If I remember correctly, the default pre 6.5.0 behaviour was that it was ignored, hence false. Just so your aware

You mean true here though right? 6.5.0 ignored self intersections yes?

**Comment by 01100100** [comment]
Let me know if you want me to implement any changes here and I would be happy to help :rescue_worker_helmet:

**Comment by smallsaucepan** [comment]
This has been achieved with the release of v7.2.0. Thanks for your help with this @01100100 and @pm0u

**Comment by 01100100** [comment]
Thank you for the great maintaining, you are both a small sauce pan and a big frying pan for this community! Cheers to you:1st_place_medal:

**Reply by smallsaucepan** [reply]
Ha ha. Thanks @01100100. Bringing all the kitchen utensils to bear.
